File: backend/src/atlas/api/routes/workload_identities.py
from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Annotated, NoReturn

from fastapi import APIRouter, Depends, Header, Query, Request, Response

from atlas.api.errors import AtlasError
from atlas.api.schemas import ResponseMeta
from atlas.api.security import (
    authorize_workload_identity_create,
    authorize_workload_identity_read,
    authorize_workload_identity_revoke,
    authorize_workload_identity_rotate,
    browser_session_subject,
)
from atlas.api.workload_identity_schemas import (
    STABLE_ID_PATTERN,
    CreateWorkloadIdentityPayload,
    CurrentWorkloadIdentityData,
    CurrentWorkloadIdentityResponse,
    IssuedWorkloadCredentialData,
    IssuedWorkloadCredentialResponse,
    RevokeWorkloadCredentialPayload,
    RotateWorkloadCredentialPayload,
    WorkloadCredentialData,
    WorkloadCredentialResponse,
    WorkloadIdentityData,
    WorkloadIdentityInventoryData,
    WorkloadIdentityInventoryResponse,
)
from atlas.modules.authorization.domain.models import AuthorizationDecision
from atlas.modules.identity.application.workload_identities import (
    WorkloadIdentityError,
    WorkloadIdentityService,
)
from atlas.modules.identity.domain.models import AuthenticatedSubject

router = APIRouter(prefix="/workload-identities", tags=["workload-identities"])
IDEMPOTENCY_HEADER = Header(
    alias="Idempotency-Key", min_length=8, max_length=128, pattern=r"^[A-Za-z0-9._:-]+$"
)


def _meta(request: Request) -> ResponseMeta:
    return ResponseMeta(
        correlation_id=str(request.state.correlation_id), generated_at=datetime.now(UTC)
    )


def _raise_workload_error(error: WorkloadIdentityError) -> NoReturn:
    if error.code == "workload_idempotency_conflict":
        raise AtlasError(
            status=409,
            code=error.code,
            title="Workload identity conflict",
            detail="The idempotency key is already bound to a different request.",
        ) from error
    if error.code in {"workload_rotation_unavailable", "workload_identity_unavailable"}:
        raise AtlasError(
            status=409,
            code="workload_identity_unavailable",
            title="Workload identity unavailable",
            detail="The requested workload identity state is unavailable.",
        ) from error
    if error.code == "workload_authentication_failed":
        raise AtlasError(
            status=401,
            code=error.code,
            title="Workload authentication failed",
            detail="The workload credential is invalid or no longer active.",
        ) from error
    if error.code == "enterprise_human_required":
        raise AtlasError(
            status=403,
            code="authorization_denied",
            title="Request denied",
            detail="Workload identity governance is not authorized.",
        ) from error
    raise AtlasError(
        status=404,
        code="workload_credential_unavailable",
        title="Workload credential unavailable",
        detail="The requested workload credential is unavailable.",
    ) from error


@router.get("", response_model=WorkloadIdentityInventoryResponse)
async def get_workload_identities(
    request: Request,
    response: Response,
    subject: Annotated[AuthenticatedSubject, Depends(browser_session_subject)],
    _decision: Annotated[AuthorizationDecision, Depends(authorize_workload_identity_read)],
    query: Annotated[str | None, Query(max_length=128)] = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
) -> WorkloadIdentityInventoryResponse:
    service: WorkloadIdentityService = request.app.state.workload_identity_service
    try:
        inventory = await service.inventory(
            actor=subject,
            query=query,
            limit=limit,
            correlation_id=str(request.state.correlation_id),
        )
    except WorkloadIdentityError as exc:
        _raise_workload_error(exc)
    response.headers["Cache-Control"] = "no-store"
    return WorkloadIdentityInventoryResponse(
        data=WorkloadIdentityInventoryData(
            identities=[WorkloadIdentityData.from_domain(item) for item in inventory.identities],
            credentials=[
                WorkloadCredentialData.from_domain(item) for item in inventory.credentials
            ],
            truncated=inventory.truncated,
        ),
        meta=_meta(request),
    )


@router.post("", response_model=IssuedWorkloadCredentialResponse, status_code=201)
async def create_workload_identity(
    payload: CreateWorkloadIdentityPayload,
    request: Request,
    response: Response,
    subject: Annotated[AuthenticatedSubject, Depends(browser_session_subject)],
    _decision: Annotated[AuthorizationDecision, Depends(authorize_workload_identity_create)],
    idempotency_key: Annotated[str, IDEMPOTENCY_HEADER],
) -> IssuedWorkloadCredentialResponse:
    service: WorkloadIdentityService = request.app.state.workload_identity_service
    try:
        issued = await service.create(
            actor=subject,
            identity_id=payload.identity_id,
            display_name=payload.display_name,
            service_id=payload.service_id,
            instance_id=payload.instance_id,
            owner_subject_id=payload.owner_subject_id,
            purpose=payload.purpose,
            audiences=tuple(payload.audiences),
            secret_reference_ids=tuple(payload.secret_reference_ids),
            lifetime=timedelta(minutes=payload.lifetime_minutes),
            reason=payload.reason,
            idempotency_key=idempotency_key,
            correlation_id=str(request.state.correlation_id),
        )
    except WorkloadIdentityError as exc:
        _raise_workload_error(exc)
    response.headers["Cache-Control"] = "no-store"
    return IssuedWorkloadCredentialResponse(
        data=IssuedWorkloadCredentialData.from_domain(issued), meta=_meta(request)
    )


@router.post(
    "/{identity_id}/rotations",
    response_model=IssuedWorkloadCredentialResponse,
    status_code=201,
)
async def rotate_workload_credential(
    identity_id: str,
    payload: RotateWorkloadCredentialPayload,
    request: Request,
    response: Response,
    subject: Annotated[AuthenticatedSubject, Depends(browser_session_subject)],
    _decision: Annotated[AuthorizationDecision, Depends(authorize_workload_identity_rotate)],
    idempotency_key: Annotated[str, IDEMPOTENCY_HEADER],
) -> IssuedWorkloadCredentialResponse:
    service: WorkloadIdentityService = request.app.state.workload_identity_service
    try:
        issued = await service.rotate(
            identity_id,
            actor=subject,
            expected_version=payload.expected_version,
            lifetime=timedelta(minutes=payload.lifetime_minutes),
            overlap=timedelta(minutes=payload.overlap_minutes),
            reason=payload.reason,
            idempotency_key=idempotency_key,
            correlation_id=str(request.state.correlation_id),
        )
    except WorkloadIdentityError as exc:
        _raise_workload_error(exc)
    response.headers["Cache-Control"] = "no-store"
    return IssuedWorkloadCredentialResponse(
        data=IssuedWorkloadCredentialData.from_domain(issued), meta=_meta(request)
    )


@router.post(
    "/credentials/{credential_id}/revocations",
    response_model=WorkloadCredentialResponse,
)
async def revoke_workload_credential(
    credential_id: str,
    payload: RevokeWorkloadCredentialPayload,
    request: Request,
    response: Response,
    subject: Annotated[AuthenticatedSubject, Depends(browser_session_subject)],
    _decision: Annotated[AuthorizationDecision, Depends(authorize_workload_identity_revoke)],
    idempotency_key: Annotated[str, IDEMPOTENCY_HEADER],
) -> WorkloadCredentialResponse:
    service: WorkloadIdentityService = request.app.state.workload_identity_service
    try:
        record = await service.revoke(
            credential_id,
            actor=subject,
            expected_version=payload.expected_version,
            reason=payload.reason,
            idempotency_key=idempotency_key,
            correlation_id=str(request.state.correlation_id),
        )
    except WorkloadIdentityError as exc:
        _raise_workload_error(exc)
    response.headers["Cache-Control"] = "no-store"
    return WorkloadCredentialResponse(
        data=WorkloadCredentialData.from_domain(record), meta=_meta(request)
    )


@router.get("/current", response_model=CurrentWorkloadIdentityResponse)
async def get_current_workload_identity(
    request: Request,
    response: Response,
    authorization: Annotated[str, Header(alias="Authorization")],
    audience: Annotated[str, Header(alias="X-Atlas-Audience", pattern=STABLE_ID_PATTERN)],
    environment_id: Annotated[str, Header(alias="X-Atlas-Environment", pattern=STABLE_ID_PATTERN)],
) -> CurrentWorkloadIdentityResponse:
    scheme, separator, token = authorization.partition(" ")
    if separator != " " or scheme.lower() != "workload":
        token = ""
    service: WorkloadIdentityService = request.app.state.workload_identity_service
    try:
        subject = await service.authenticate(
            token,
            audience=audience,
            environment_id=environment_id,
            correlation_id=str(request.state.correlation_id),
        )
    except WorkloadIdentityError as exc:
        _raise_workload_error(exc)
    response.headers["Cache-Control"] = "no-store"
    return CurrentWorkloadIdentityResponse(
        data=CurrentWorkloadIdentityData.from_domain(
            subject, audience=audience, environment_id=environment_id
        ),
        meta=_meta(request),
    )

File: frontend/src/api/workloadIdentities.ts
import { apiFetch, ApiRequestError } from "./client";

export type WorkloadIdentity = {
  identity_id: string;
  version: number;
  display_name: string;
  service_id: string;
  instance_id: string;
  owner_subject_id: string;
  purpose: string;
  organization_id: string;
  environment_id: string;
  audiences: string[];
  secret_reference_ids: string[];
  state: "active" | "disabled";
  created_at: string;
  updated_at: string;
};

export type WorkloadCredential = {
  credential_id: string;
  version: number;
  identity_id: string;
  key_version: number;
  audiences: string[];
  issued_at: string;
  expires_at: string;
  state: "active" | "retiring" | "revoked" | "expired";
  retire_at: string | null;
  revoked_at: string | null;
};

export type WorkloadIdentityInventoryResponse = {
  data: {
    identities: WorkloadIdentity[];
    credentials: WorkloadCredential[];
    truncated: boolean;
  };
};

export type IssuedWorkloadCredentialResponse = {
  data: {
    identity: WorkloadIdentity;
    credential: WorkloadCredential;
    token: string;
  };
};

export async function getWorkloadIdentities(
  query: string,
): Promise<WorkloadIdentityInventoryResponse | null> {
  const parameters = new URLSearchParams({ limit: "50" });
  if (query.trim()) parameters.set("query", query.trim());
  const response = await apiFetch(`/api/v1/workload-identities?${parameters.toString()}`, {
    headers: { Accept: "application/json" },
  });
  if (response.status === 403) return null;
  if (!response.ok) throw new ApiRequestError("Workload identity inventory failed", response.status);
  const payload = (await response.json()) as WorkloadIdentityInventoryResponse;
  if (!Array.isArray(payload.data?.identities) || !Array.isArray(payload.data?.credentials)) {
    throw new ApiRequestError("Workload identity inventory failed", response.status);
  }
  return payload;
}

export async function createWorkloadIdentity(input: {
  identityId: string;
  displayName: string;
  serviceId: string;
  instanceId: string;
  ownerSubjectId: string;
  purpose: string;
  audience: string;
  secretReferenceId: string;
  lifetimeMinutes: number;
  reason: string;
  idempotencyKey: string;
}): Promise<IssuedWorkloadCredentialResponse> {
  const response = await apiFetch("/api/v1/workload-identities", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "Idempotency-Key": input.idempotencyKey,
    },
    body: JSON.stringify({
      identity_id: input.identityId,
      display_name: input.displayName,
      service_id: input.serviceId,
      instance_id: input.instanceId,
      owner_subject_id: input.ownerSubjectId,
      purpose: input.purpose,
      audiences: [input.audience],
      secret_reference_ids: [input.secretReferenceId],
      lifetime_minutes: input.lifetimeMinutes,
      reason: input.reason,
    }),
  });
  if (!response.ok) throw new ApiRequestError("Workload identity creation failed", response.status);
  return (await response.json()) as IssuedWorkloadCredentialResponse;
}

export async function rotateWorkloadCredential(input: {
  identityId: string;
  expectedVersion: number;
  lifetimeMinutes: number;
  overlapMinutes: number;
  reason: string;
  idempotencyKey: string;
}): Promise<IssuedWorkloadCredentialResponse> {
  const response = await apiFetch(
    `/api/v1/workload-identities/${encodeURIComponent(input.identityId)}/rotations`,
    {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "Idempotency-Key": input.idempotencyKey,
      },
      body: JSON.stringify({
        expected_version: input.expectedVersion,
        lifetime_minutes: input.lifetimeMinutes,
        overlap_minutes: input.overlapMinutes,
        reason: input.reason,
      }),
    },
  );
  if (!response.ok) throw new ApiRequestError("Workload credential rotation failed", response.status);
  return (await response.json()) as IssuedWorkloadCredentialResponse;
}

export async function revokeWorkloadCredential(input: {
  credentialId: string;
  expectedVersion: number;
  reason: string;
  idempotencyKey: string;
}): Promise<void> {
  const response = await apiFetch(
    `/api/v1/workload-identities/credentials/${encodeURIComponent(input.credentialId)}/revocations`,
    {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "Idempotency-Key": input.idempotencyKey,
      },
      body: JSON.stringify({
        expected_version: input.expectedVersion,
        reason: input.reason,
      }),
    },
  );
  if (!response.ok) throw new ApiRequestError("Workload credential revocation failed", response.status);
}
